### DISCUSSION TITLE
Trying to Publish on a Topic to Tasmota with MQTT

### DISCUSSION THREAD

**Discussion by jackpotcityco** [arendst/Tasmota#19734] [Q&A] [answered]
Hello,

**Program Version 13.1.0.4(tasmota)
Build Date & Time 2023-10-09T14:35:26
Core/SDK Version 2_7_4_9/2.2.2-dev(38a443e)**

I am using:
**ESP8266EX
VL53L1X - Time-of-Flight (ToF) ranging sensor**

I am trying to Publish a topic from a MQTT client in C# to Tasmota in order to activate `Rule2` but do not succeed to do this. (This is my first time I try to do this). Tasmota MQTT is connected to Mosquitto successfully.

In the documentation I could read the below:
```
Commands over MQTT are issued to Tasmota by using topic cmnd/%topic%/<command> and payload <parameter>. If there is no <parameter> (an empty MQTT message/payload),
//a query is sent for current status of the <command>
```
My configuration for the Topic and Full Topic is as below:
![image](https://github.com/arendst/Tasmota/assets/47510843/0e606a57-80a0-4890-a73d-8e86098796f7)

I have tried to put all those 3 combinations as a test in the C# code where none of them seems to Publish to Tasmota web console:
```
            string topic = "cmnd/device/tasmota_01/function1";
            string messagePayload = "function1=trigger"; // Include a payload without a specific value

            string topic = "device/tasmota_01/function1";
            string messagePayload = "function1=trigger"; // Include a payload without a specific value

            string topic = "tasmota_01/function1";
            string messagePayload = "function1=trigger"; // Include a payload without a specific value
```
Full C# code:
```
            MqttClient client = new MqttClient("15.1.0.17"); // Create the client
            client.Connect("PC_client"); // Connect to the MQTT broker
            string topic = "tasmota_01/function1";  // Publish a message to a topic
            string messagePayload = "function1=trigger"; // Include a payload without a specific value
            client.Publish(topic, System.Text.Encoding.UTF8.GetBytes(messagePayload), MqttMsgBase.QOS_LEVEL_AT_LEAST_ONCE, false);
            client.Disconnect(); // Disconnect from the MQTT broker
```

The Mosquitto broker responds like this when I publish:
```
1697234464: New connection from 15.1.0.17:58566 on port 1883.
1697234464: New client connected from 15.1.0.17:58566 as PC_client (p2, c1, k60).
1697234464: No will message specified.
1697234464: Sending CONNACK to PC_client (0, 0)
1697234464: Received DISCONNECT from PC_client
1697234464: Client PC_client disconnected.
```

The Rule2 that I have put in the Tasmota Console as a test:
Rule2 ON tasmota_01#function1 DO ConsoleLog, Distancecheck ENDON
![image](https://github.com/arendst/Tasmota/assets/47510843/ba5de754-3356-430b-bc0c-fa25ecef5261)

When I Publish with the C# code, the goal would be for Tasmota to output the string: "Distancecheck" in the console but this do not happen. I am not sure if it is possible so see if I could miss something?

Thank you!

**Comment by sfromis** [comment]
This is not how it works. Basically, you cannot create your own commands with Tasmota on ESP8266. And rule triggers are not acting on incoming MQTT payloads.

To have Tasmota react to such topics, do not publish them to the `cmnd` prefix, have the payload in JSON format, and use the [subscribe](https://tasmota.github.io/docs/MQTT/#subscribeunsubscribe) feature to pick up a single value from the JSON.

A much easier way of doing it would be to publish an Event command, and then you can have triggers picking up this. Something like
topic `cmnd/device/tasmota_01/Event` and payload `function1=trigger`
and rule `on Event#function1=trigger do ConsoleLog, Distancecheck endon`

Of course, this will give an error message as `ConsoleLog` is not a valid command, fix that!

If you used ESP32, you'd have much better features for handling MQTT subscriptions via the Berry language.

When it comes to your C# coding, note that the MQTT library used by Tasmota does not support QoS > 0.

**Reply by jackpotcityco** [reply]
> A much easier way of doing it would be to publish an Event command, and then you can have triggers picking up this. Something like
> topic cmnd/device/tasmota_01/Event and payload function1=trigger
> and rule on Event#function1=trigger do ConsoleLog, Distancecheck endon

Sorry for my beginner questions but I am not sure if the above was an alternative instead of using JSON and `subscribe` as that would need to add "SUPPORT_MQTT_EVENT" which I am not familiar where to do. I am using a C# client at this time.
I can use: `MqttMsgBase.QOS_LEVEL_AT_MOST_ONCE` as Tasmota does not support QoS > 0.

I am not sure if I understood this correct. But I did put this in the C# code:
```
            string topic = "cmnd/device/tasmota_01/Event";
            string messagePayload = "function1=trigger"; // Include a payload without a specific value
            client.Publish(topic, System.Text.Encoding.UTF8.GetBytes(messagePayload), MqttMsgBase.QOS_LEVEL_AT_MOST_ONCE, false);
```

I then put this in the Tasmota console. I changed `ConsoleLog` to `Console`. Rule2 State is `ON`.
`Rule2 ON Event#function1=trigger DO Console,Distancecheck ENDON`

However, Tasmota does not react to that C# code in the Console?

**Reply by sfromis** [reply]
Using `Event` was indeed **instead** of the laborious approach using the subscribe feature.

And why are you keeping using commands Tasmota does not have? There is no "Console" command either, and nothing similar except for what I already suggested earlier.. Still, you'd see messages getting to the bad command, so this is not an explanation for getting nothing.

I'd suggest verifying details, Tasmota should display messages about what topics it subscribes to when restarting (and having `Weblog 3` set for details), messages like `MQT: Subscribe to ....`. Also verify rule present and activated using the command `Rule0`.

**Comment by barbudor** [comment, marked as answer]
```haskell
09:05:51.785 CMD: Weblog 3
09:05:51.793 MQT: stat/ble-bridge/WEBLOG = {"WebLog":3}
09:05:52.774 CFG: Saved, Count 547, Bytes 4096
09:06:40.889 CMD: Rule2 ON Event#function1=trigger DO Console,Distancecheck ENDON
09:06:40.892 SRC: WebConsole from 192.168.168.50
09:06:40.894 CMD: Grp 0, Cmd 'RULE', Idx 2, Len 57, Pld -99, Data 'ON Event#function1=trigger DO Console,Distancecheck ENDON'
09:06:40.896 RUL: Stored uncompressed, would compress from 57 to 43 (-25%)
09:06:40.904 MQT: stat/ble-bridge/RULE = {"Rule2":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":57,"Free":454,"Rules":"ON Event#function1=trigger DO Console,Distancecheck ENDON"}}
```
Sending from pc
Topic:        `/cmnd/ble-bridge/event`
Payload:     `function1=trigger`
```haskell
09:06:48.428 SRC: MQTT
09:06:48.431 CMD: Grp 0, Cmd 'EVENT', Idx 1, Len 17, Pld -99, Data 'function1=trigger'
09:06:48.438 MQT: stat/ble-bridge/EVENT = {"Event":"Done"}
09:06:48.456 RUL: EVENT#FUNCTION1=TRIGGER performs "Console,Distancecheck"
09:06:48.457 SRC: Rule
09:06:48.459 CMD: Grp 0, Cmd 'CONSOLE', Idx 1, Len 14, Pld -99, Data ',Distancecheck'
09:06:48.466 MQT: stat/ble-bridge/COMMAND = {"Command":"Unknown"}
```

**Reply by barbudor** [reply]
Note that in my exemple, the topic is `cmnd/ble-bridge/event` because my Tasmota's topic is `ble-bridge`
If you have setup your Tasmota's topic to `device/tasmota_01` then the topic for the event command is `cmnd/device/tasmota_01/event`

**Comment by jackpotcityco** [comment]
Thank you for that information.

I managed to find out the problem with your help. It seems that when I Disconnected with the C# code, the broker didn't have time to push this through to Tasmota in this case. I was thinking that all code before this line should have gone through. But when I removed the line below. It was showed in Tasmota, going through the broker:
`client.Disconnect();` // Disconnect from the MQTT broker

That was great to use `Weblog 3` and below was seen in Tasmota where it shows it subscribes on all Topics "#" under for example: `cmnd/device/tasmota_01/#`
```
17:54:21.145 MQT: Attempting connection...
17:54:21.245 MQT: Connected
17:54:21.250 MQT: tele/device/tasmota_01/LWT = Online (retained)
17:54:21.253 MQT: cmnd/device/tasmota_01/POWER = 
17:54:21.255 MQT: Subscribe to cmnd/device/tasmota_01/#
17:54:21.258 MQT: Subscribe to cmnd/tasmotas/#
17:54:21.260 MQT: Subscribe to cmnd/DVES_18076C_fb/#
```
Yes, I am not sure why I use the `Console/ConsoleLog` command. I saw that in an example somewhere.
Is there a command for writing to the console. I am not sure if I can find this in the documentation?

Thank you for your help!

**Reply by sfromis** [reply]
As I already responded..... https://github.com/arendst/Tasmota/discussions/19723#discussioncomment-7265477

**Reply by barbudor** [reply]
There is no command specifically to display something in the console
What sfromis suggest is the commonly used method of setting a variable to a message as a way to create some logs

**Reply by jackpotcityco** [reply]
Yes, now I understand. I used "var1" and it outputs a nice Json format like below instead and `{"Command":"Unknown"}` is no longer seen:

`20:57:15.341 MQT: stat/device/tasmota_01/RESULT = {"Var1":"DistanceB,30.5"}`

Thank you!
